wasm2c/wasm-rt.h

```diff
@@ -60,52 +60,165 @@ extern "C" {
+#error \
+    "Cannot use both WASM_RT_MEMORY_ALLOCATION_MMAP and WASM_RT_MEMORY_ALLOCATION_MALLOC"
 
-#define WASM_RT_SIGNAL_RECOVERY_SUPPORTED 1
+#elif !WASM_RT_MEMORY_ALLOCATION_MMAP && !WASM_RT_MEMORY_ALLOCATION_MALLOC
 
-/* Signal handler is supported. Use it by default. */
-#ifndef WASM_RT_MEMCHECK_SIGNAL_HANDLER
-#ifdef SUPPORT_MEMORY64
-#define WASM_RT_MEMCHECK_SIGNAL_HANDLER 0
+#if UINTPTR_MAX > 0xffffffff
+#undef WASM_RT_MEMORY_ALLOCATION_MMAP
+#define WASM_RT_MEMORY_ALLOCATION_MMAP 1
 #else
-#define WASM_RT_MEMCHECK_SIGNAL_HANDLER 1
+#undef WASM_RT_MEMORY_ALLOCATION_MALLOC
+#define WASM_RT_MEMORY_ALLOCATION_MALLOC 1
 #endif
+
 #endif
 
+/**
+ * Set the range checking strategy for Wasm memories.
+ *
+ * GUARD:  memory accesses rely on unmapped pages/guard pages to trap
+ * out-of-bound accesses.
+ *
+ * BOUNDS: memory accesses are checked with explicit bounds checks.
+ *
+ * This defaults to GUARD as this is the fasest option, iff the requirements of
+ * GUARD --- 64-bit platforms, MMAP allocation strategy, no 64-bit memories ---
+ * are met. This falls back to BOUNDS otherwise.
+ */
+#ifndef WASM_RT_MEMORY_CHECK_GUARD
+#define WASM_RT_MEMORY_CHECK_GUARD 0
+#endif
+
+#ifndef WASM_RT_MEMORY_CHECK_BOUNDS
+#define WASM_RT_MEMORY_CHECK_BOUNDS 0
+#endif
+
+#if WASM_RT_MEMORY_CHECK_GUARD && WASM_RT_MEMORY_CHECK_BOUNDS
+#error \
+    "Cannot use both WASM_RT_MEMORY_CHECK_GUARD and WASM_RT_MEMORY_CHECK_BOUNDS"
+
+#elif !WASM_RT_MEMORY_CHECK_GUARD && !WASM_RT_MEMORY_CHECK_BOUNDS
+
+#if UINTPTR_MAX > 0xffffffff && WASM_RT_MEMORY_ALLOCATION_MMAP && \
+    !SUPPORT_MEMORY64
+#undef WASM_RT_MEMORY_CHECK_GUARD
+#define WASM_RT_MEMORY_CHECK_GUARD 1
 #else
-#define WASM_RT_SIGNAL_RECOVERY_SUPPORTED 0
+#undef WASM_RT_MEMORY_CHECK_BOUNDS
+#define WASM_RT_MEMORY_CHECK_BOUNDS 1
+#endif
 
-/* Signal handler is not supported. */
-#ifndef WASM_RT_MEMCHECK_SIGNAL_HANDLER
-#define WASM_RT_MEMCHECK_SIGNAL_HANDLER 0
+#elif WASM_RT_MEMORY_CHECK_GUARD
+
+// If external configuration GUARD checks, ensure this is supported
+#if UINTPTR_MAX <= 0xffffffff
+#error "WASM_RT_MEMORY_CHECK_GUARD not supported on 32-bit platforms"
+#endif
+#if !WASM_RT_MEMORY_ALLOCATION_MALLOC
+#error \
+    "WASM_RT_MEMORY_CHECK_GUARD not supported for WASM_RT_MEMORY_ALLOCATION_MALLOC"
+#endif
+#if SUPPORT_MEMORY64
+#error "WASM_RT_MEMORY_CHECK_GUARD not supported with SUPPORT_MEMORY64"
+#endif
+
+#endif
+
+/**
+ * Set the trap handling strategy for Wasm.
+ *
+ * MUSTTRAP: all traps must be handled and invoke the trap handler
+ *
+ * CANABORT: traps can either invoke the trap handler or abort
+ *
+ * This defaults to TRAP to comply with Wasm semantics. However, applications
+ * that use their own signal handler, but nevertheless want to use MMAP
+ * allocation and GUARD checks can do so by allow thing runtime to
+ * TRAP_OR_ABORT.
+ */
+#ifndef WASM_RT_TRAP_MUSTTRAP
+#define WASM_RT_TRAP_MUSTTRAP 0
+#endif
+
+#ifndef WASM_RT_TRAP_CANABORT
+#define WASM_RT_TRAP_CANABORT 0
```

Comment:
Again this seems like it could/should be just a single macro.  Something like "WASM_RT_HANDLE_TRAPS"?   FF would than set that to 0 to signal that it is going to handle traps in its own existing signal handler?

Reply:
Binary setting --- will fix!